core/state/state_witness.go

```diff
@@ -0,0 +1,275 @@
+package state
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/rawdb"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/rlp"
+	"os"
+)
+
+type Witness struct {
+	block       *types.Block
+	blockHashes map[uint64]common.Hash
+	codes       map[common.Hash]Code
+	root        common.Hash
+	lists       map[common.Hash]map[string][]byte
+}
+
+func (w *Witness) GetBlockHash(num uint64) common.Hash {
+	return w.blockHashes[num]
+}
+
+func (w *Witness) Root() common.Hash {
+	return w.root
+}
+
+type encodedWitness struct {
+	block       types.Block
+	root        common.Hash
+	owners      []common.Hash
+	allPaths    [][]string
+	allNodes    [][][]byte
+	blockNums   []uint64
+	blockHashes []common.Hash
+	codes       []Code
+	codeHashes  []common.Hash
+}
+
+func (e *encodedWitness) ToWitness() *Witness {
+	res := NewWitness()
+	res.root = e.root
+	for i := 0; i < len(e.codes); i++ {
+		res.codes[e.codeHashes[i]] = e.codes[i]
+	}
+	for i, owner := range e.owners {
+		pathMap := make(map[string][]byte)
+		for j := 0; j < len(e.allPaths[i]); j++ {
+			pathMap[e.allPaths[i][j]] = e.allNodes[i][j]
+		}
+		res.lists[owner] = pathMap
+	}
+	for i, blockNum := range e.blockNums {
+		res.blockHashes[blockNum] = e.blockHashes[i]
+	}
+	return res
+}
+
+func DecodeWitnessRLP(b []byte) (*types.Block, *Witness, error) {
+	var res encodedWitness
+	stream := rlp.NewStream(bytes.NewBuffer(b), 1_000_000)
+	_, err := stream.List()
+	if err != nil {
+		panic(err)
+	}
+	err = res.block.DecodeRLP(stream)
+	if err != nil {
+		panic(err)
+	}
+	if err = stream.Decode(&res.root); err != nil {
+		panic(err)
+	}
+	if err = stream.Decode(&res.owners); err != nil {
+		panic(err)
+	}
+	if err = stream.Decode(&res.allPaths); err != nil {
+		panic(err)
+	}
+	if err = stream.Decode(&res.allNodes); err != nil {
+		panic(err)
+	}
+	if err = stream.Decode(&res.blockNums); err != nil {
+		panic(err)
+	}
+	if err = stream.Decode(&res.blockHashes); err != nil {
+		panic(err)
+	}
+	if err = stream.Decode(&res.codeHashes); err != nil {
+		panic(err)
+	}
+	if err = stream.Decode(&res.codes); err != nil {
+		panic(err)
+	}
+	return &res.block, res.ToWitness(), nil
+}
+
+func (w *Witness) EncodeRLP() []byte {
```

Comment:
Skip this entire method, let rlp module encode it (like above)

Reply:
encoding maps isn't supported by our rlp encoder so I have to keep the internal encoding type and provide encoding/decoding methods.

Reply:
Optionally, we could add the behavior to encode maps like:
```
[
    [ keys... ],
    [ values... ],
]```

Reply:
See https://github.com/ethereum/go-ethereum/pull/28640#discussion_r1424360490

Reply:
Sure.  The downside of this is that we will need to expose two types `Witness`/`RLPWitness`.  Which is fine I guess.

Reply:
Actually we can just keep the `rlpWitness` private.  nevermind.